src/Compiler/Utilities/Cancellable.fs

```diff
@@ -92,14 +96,19 @@ module Cancellable =
 
     let toAsync c =
         async {
+            use! _holder = Cancellable.UseToken()
+
             let! ct = Async.CancellationToken
-            let res = run ct c
 
             return!
-                Async.FromContinuations(fun (cont, _econt, ccont) ->
-                    match res with
-                    | ValueOrCancelled.Value v -> cont v
-                    | ValueOrCancelled.Cancelled ce -> ccont ce)
+                Async.FromContinuations(fun (cont, econt, ccont) ->
+                    try
+                        match run ct c with
+                        | ValueOrCancelled.Value v -> cont v
+                        | ValueOrCancelled.Cancelled ce -> ccont ce
+                    with
+                    | :? OperationCanceledException as ce when ct.IsCancellationRequested -> ccont ce
+                    | :? OperationCanceledException as e -> InvalidOperationException("Wrong cancellation token", e) |> econt)
```

Comment:
The purpose of the last handler clause in `with` was to catch programming errors early (e.g. tokens getting mixed up).

Is this now really executed at all high-level endpoints at least, to keep to safety guarantees at similar level?

This assumes that eventually, at least on the public API surface, every cancellable once becomes an async via `toAsync`, correct?

Reply:
Well, there is also `Cancellable.runWithoutCancellation` on the API surface which I think is in use only in non-cancellable contexts (fsc and fsi). With this PR it will just let any potential OCE pass through like any other uncaught exception.

Apart from this, yes, I think currently each cancellable eventually leads to an async service endpoint via `toAsync`, from which it gets its cancellation token.

Reply:
That is good, we do not need the special detection in `runWithoutCancellation` anyway.
LGTM